src/instrument.cpp

```diff
@@ -613,16 +1031,29 @@ void JNICALL Instrument::ClassFileLoadHook(jvmtiEnv* jvmti, JNIEnv* jni,
 
     bool wildcard_class = _target_class[0] == '*' && strlen(_target_class) == 1;
     if (name == NULL || wildcard_class || strcmp(name, _target_class) == 0) {
-        BytecodeRewriter rewriter(class_data, class_data_len, _target_class);
+        if (wildcard_class && strncmp("one/profiler/", name, 13) == 0) {
+            return;
+        }
+        BytecodeRewriter rewriter(class_data, class_data_len, _target_class, _latency >= 0);
         rewriter.rewrite(new_class_data, new_class_data_len);
     }
 }
 
-void JNICALL Instrument::recordSample(JNIEnv* jni, jobject unused) {
+void JNICALL Instrument::recordEntry(JNIEnv* jni, jobject unused) {
     if (!_enabled) return;
 
     if (_interval <= 1 || ((atomicInc(_calls) + 1) % _interval) == 0) {
         ExecutionEvent event(TSC::ticks());
         Profiler::instance()->recordSample(NULL, _interval, INSTRUMENTED_METHOD, &event);
     }
 }
+
+void JNICALL Instrument::recordExit(JNIEnv* jni, jobject unused, jlong startTimeNanos) {
+    if (!_enabled) return;
+
+    u64 duration = OS::nanotime() - (u64) startTimeNanos;
+    if (duration >= _latency) {
+        ExecutionEvent event((u64) startTimeNanos * TSC::frequency() / NANOTIME_FREQ);
```

Comment:
Should we protect against overflow here? We may lose some precision by rearranging the operands

Reply:
The forumla is incorrect: TSC clock and monotonic clock have different bases.
And yes, it's better to use `double` arithmetic to protect from overflow.

Reply:
Okay, just to make sure I got it right: I'm going to do something like this:
```
jmethodID counterTime = env->GetStaticMethodID(cls, "counterTime", "()J");
u64 jvm_ticks = env->CallStaticLongMethod(cls, counterTime);

u64 nnt_offset_ticks = OS::nanotime() * TSC::frequency() - jvm_ticks;
u64 start_time_ticks = (double) startTimeNanos * TSC::frequency() / NANOTIME_FREQ - nnt_offset_ticks;
```

Reply:
Got it: f263b4f1b318e3294da902c3b7611688fe857b99

> make sure OS::nanotime and TSC::ticks are called as close to each other as possible.

There's an `if` in between, not sure if that hurts more than calling `TSC::ticks` regardless